From 100251b8221f88cccb27ad3997a7010a77501a44 Mon Sep 17 00:00:00 2001
From: Valentine Barshak <valentine.barshak@cogentembedded.com>
Date: Mon, 17 Jun 2019 18:15:58 +0300
Subject: [PATCH 5/5] media: i2c: soc_camera: Fix more "Bad of_node_put" errors

This fixes "OF: ERROR: Bad of_node_put()" error and possible kernel crash
when probing i2c soc_camera sensor devices in the recently added drivers.

When CONFIG_OF_DYNAMIC is not enabled, the error message is not displayed,
and everything seems to work fine. However, there's an error in the sensor
parse_dt callbacks which causes inconsistent reference counter management.

Move of_node_put() out of a loop scope to prevent bad of_node_put error.
We don't need to call of_node_put() for every node since subsequent
calls to of_graph_get_next_endpoint() decrement reference counter of
the previous node. We only need to call of_node_put() for the last node.

Signed-off-by: Valentine Barshak <valentine.barshak@cogentembedded.com>
---
 drivers/media/i2c/soc_camera/ar0323.c  | 4 ++--
 drivers/media/i2c/soc_camera/isx016.c  | 4 ++--
 drivers/media/i2c/soc_camera/isx019.c  | 4 ++--
 drivers/media/i2c/soc_camera/ov10640.c | 4 ++--
 drivers/media/i2c/soc_camera/ov2311.c  | 4 ++--
 5 files changed, 10 insertions(+), 10 deletions(-)

diff --git a/drivers/media/i2c/soc_camera/ar0323.c b/drivers/media/i2c/soc_camera/ar0323.c
index 2ee5526..2104d6a 100644
--- a/drivers/media/i2c/soc_camera/ar0323.c
+++ b/drivers/media/i2c/soc_camera/ar0323.c
@@ -425,8 +425,6 @@ static int ar0323_parse_dt(struct device_node *np, struct ar0323_priv *priv)
 		if (!endpoint)
 			break;
 
-		of_node_put(endpoint);
-
 		rendpoint = of_parse_phandle(endpoint, "remote-endpoint", 0);
 		if (!rendpoint)
 			continue;
@@ -438,6 +436,8 @@ static int ar0323_parse_dt(struct device_node *np, struct ar0323_priv *priv)
 			break;
 	}
 
+	of_node_put(endpoint);
+
 	if (!priv->ti9x4_addr) {
 		dev_err(&client->dev, "deserializer does not present\n");
 		return -EINVAL;
diff --git a/drivers/media/i2c/soc_camera/isx016.c b/drivers/media/i2c/soc_camera/isx016.c
index eab155c..ab85b7d 100644
--- a/drivers/media/i2c/soc_camera/isx016.c
+++ b/drivers/media/i2c/soc_camera/isx016.c
@@ -425,8 +425,6 @@ static int isx016_parse_dt(struct device_node *np, struct isx016_priv *priv)
 		if (!endpoint)
 			break;
 
-		of_node_put(endpoint);
-
 		rendpoint = of_parse_phandle(endpoint, "remote-endpoint", 0);
 		if (!rendpoint)
 			continue;
@@ -443,6 +441,8 @@ static int isx016_parse_dt(struct device_node *np, struct isx016_priv *priv)
 			break;
 	}
 
+	of_node_put(endpoint);
+
 	if (!priv->max9286_addr && !priv->ti9x4_addr) {
 		dev_err(&client->dev, "deserializer does not present for ISX016\n");
 		return -EINVAL;
diff --git a/drivers/media/i2c/soc_camera/isx019.c b/drivers/media/i2c/soc_camera/isx019.c
index c2f5a4a..188fd28 100644
--- a/drivers/media/i2c/soc_camera/isx019.c
+++ b/drivers/media/i2c/soc_camera/isx019.c
@@ -483,8 +483,6 @@ static int isx019_parse_dt(struct device_node *np, struct isx019_priv *priv)
 		if (!endpoint)
 			break;
 
-		of_node_put(endpoint);
-
 		rendpoint = of_parse_phandle(endpoint, "remote-endpoint", 0);
 		if (!rendpoint)
 			continue;
@@ -495,6 +493,8 @@ static int isx019_parse_dt(struct device_node *np, struct isx019_priv *priv)
 			break;
 	}
 
+	of_node_put(endpoint);
+
 	if (!priv->max9286_addr) {
 		dev_err(&client->dev, "deserializer does not present for ISX019\n");
 		return -EINVAL;
diff --git a/drivers/media/i2c/soc_camera/ov10640.c b/drivers/media/i2c/soc_camera/ov10640.c
index 8746988..31117e5 100644
--- a/drivers/media/i2c/soc_camera/ov10640.c
+++ b/drivers/media/i2c/soc_camera/ov10640.c
@@ -496,8 +496,6 @@ static int ov10640_parse_dt(struct device_node *np, struct ov10640_priv *priv)
 		if (!endpoint)
 			break;
 
-		of_node_put(endpoint);
-
 		of_property_read_u32(endpoint, "dvp-order", &priv->dvp_order);
 
 		rendpoint = of_parse_phandle(endpoint, "remote-endpoint", 0);
@@ -516,6 +514,8 @@ static int ov10640_parse_dt(struct device_node *np, struct ov10640_priv *priv)
 			break;
 	}
 
+	of_node_put(endpoint);
+
 	if (!priv->max9286_addr && !priv->ti9x4_addr) {
 		dev_err(&client->dev, "deserializer does not present for OV10640\n");
 		return -EINVAL;
diff --git a/drivers/media/i2c/soc_camera/ov2311.c b/drivers/media/i2c/soc_camera/ov2311.c
index 06e57dd..c8d260c 100644
--- a/drivers/media/i2c/soc_camera/ov2311.c
+++ b/drivers/media/i2c/soc_camera/ov2311.c
@@ -453,8 +453,6 @@ static int ov2311_parse_dt(struct device_node *np, struct ov2311_priv *priv)
 		if (!endpoint)
 			break;
 
-		of_node_put(endpoint);
-
 		of_property_read_u32(endpoint, "dvp-order", &priv->dvp_order);
 
 		rendpoint = of_parse_phandle(endpoint, "remote-endpoint", 0);
@@ -473,6 +471,8 @@ static int ov2311_parse_dt(struct device_node *np, struct ov2311_priv *priv)
 			break;
 	}
 
+	of_node_put(endpoint);
+
 	if (!priv->max9286_addr && !priv->ti9x4_addr) {
 		dev_err(&client->dev, "deserializer does not present for OV2311\n");
 		return -EINVAL;
-- 
2.7.4

